File: client/lib/types/credential.dart
import 'package:selfpass_protobuf/credentials.pb.dart' as protobuf;

class Metadata {
  String id;
  String sourceHost;
  DateTime createdAt;
  DateTime updatedAt;
  String primary;
  String loginUrl;
  String tag;

  Metadata({
    this.id,
    this.sourceHost,
    this.createdAt,
    this.updatedAt,
    this.primary,
    this.loginUrl,
    this.tag,
  });

  Metadata.fromProtobuf(protobuf.Metadata metadata) {
    id = metadata.id;
    createdAt = metadata.createdAt.toDateTime();
    updatedAt = metadata.updatedAt.toDateTime();
    sourceHost = metadata.sourceHost;
    primary = metadata.primary;
    loginUrl = metadata.loginUrl;
    tag = metadata.tag;
  }

  @override
  String toString() => "id: $id";
}

class MetadataInput {
  String sourceHost;
  String primary;
  String loginUrl;
  String tag;

  MetadataInput({this.sourceHost, this.primary, this.loginUrl, this.tag});
}

class Credential {
  Metadata meta;
  String username;
  String email;
  String password;
  String otpSecret;

  Credential({
    this.meta,
    this.username,
    this.email,
    this.password,
    this.otpSecret,
  });

  Credential.fromProtobuf(protobuf.Credential credential) {
    meta = Metadata(
      id: credential.id,
      createdAt: credential.createdAt.toDateTime(),
      updatedAt: credential.updatedAt.toDateTime(),
      sourceHost: credential.sourceHost,
      primary: credential.primary,
      loginUrl: credential.loginUrl,
      tag: credential.tag,
    );
    username = credential.username;
    email = credential.email;
    password = credential.password;
    otpSecret = credential.otpSecret;
  }

  @override
  String toString() => "meta: $meta\n";
}

class CredentialInput {
  MetadataInput meta;
  String username;
  String email;
  String password;
  String otpSecret;

  CredentialInput({
    this.meta,
    this.username,
    this.email,
    this.password,
    this.otpSecret,
  });
}

File: services/credentials/types/credential.go
package types

import (
	"fmt"
	"time"
)

const KeyCredential = "cred"

type Credential struct {
	Metadata
	Username  string
	Email     string
	Password  string `json:"-"`
	OTPSecret string `json:"-"`
}

func (c Credential) String() string {
	format := "%s"
	args := []interface{}{c.Metadata}

	if c.Username != "" {
		format += "username = %s\n"
		args = append(args, c.Username)
	}

	if c.Email != "" {
		format += "email = %s\n"
		args = append(args, c.Email)
	}

	return fmt.Sprintf(format, args...)
}

type CredentialInput struct {
	MetadataInput
	Username  string
	Email     string
	Password  string
	OTPSecret string
}

type Metadata struct {
	ID         string // primary key
	SourceHost string // sort key
	CreatedAt  time.Time
	UpdatedAt  time.Time
	Primary    string
	LoginURL   string
	Tag        string
}

func (m Metadata) String() string {
	format := "id = %s\ncreatedAt = %s\nupdatedAt = %s\nsourceHost = %s\nprimary = %s\n"
	args := []interface{}{m.ID, m.CreatedAt, m.UpdatedAt, m.SourceHost, m.Primary}

	if m.LoginURL != "" {
		format += "loginUrl = %s\n"
		args = append(args, m.LoginURL)
	}

	if m.Tag != "" {
		format += "tag = %s\n"
		args = append(args, m.Tag)
	}

	return fmt.Sprintf(format, args...)
}

type MetadataInput struct {
	Primary    string
	SourceHost string
	LoginURL   string
	Tag        string
}
